src/main/java/com/hazelcast/aws/impl/AwsOperation.java

```diff
@@ -0,0 +1,309 @@
+/*
+ * Copyright (c) 2008-2018, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.aws.impl;
+
+import com.hazelcast.aws.AwsConfig;
+import com.hazelcast.aws.exception.AwsConnectionException;
+import com.hazelcast.aws.security.Aws4RequestSigner;
+import com.hazelcast.aws.security.Aws4RequestSignerImpl;
+import com.hazelcast.aws.utility.Aws4RequestSignerUtils;
+import com.hazelcast.aws.security.AwsCredentials;
+import com.hazelcast.aws.utility.Environment;
+import com.hazelcast.aws.utility.MetadataUtils;
+import com.hazelcast.aws.utility.RetryUtils;
+import com.hazelcast.aws.utility.StringUtils;
+import com.hazelcast.config.InvalidConfigurationException;
+import com.hazelcast.internal.json.Json;
+import com.hazelcast.internal.json.JsonObject;
+import com.hazelcast.logging.ILogger;
+import com.hazelcast.logging.Logger;
+
+import java.io.BufferedWriter;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.io.OutputStreamWriter;
+import java.net.HttpURLConnection;
+import java.net.URL;
+import java.text.SimpleDateFormat;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Scanner;
+import java.util.TimeZone;
+import java.util.concurrent.Callable;
+import java.util.concurrent.TimeUnit;
+
+import static com.hazelcast.aws.utility.StringUtils.isEmpty;
+import static com.hazelcast.aws.utility.StringUtils.isNotEmpty;
+import static com.hazelcast.nio.IOUtil.closeResource;
+
+/**
+ * Abstract base class for AWS operations returning a response object of type E.
+ * Used by AwsClientStrategy implementations for calling AWS service endpoints.
+ */
+public abstract class AwsOperation<E> {
```

Comment:
This class (hierarchy) has a number of problems:
- It's very generic, one public method that takes some number of some objects and returns something
- Its abstract methods do not have a clear, well-defined contract (if you were to implement a new subclass, how would you know what `prepareHttpRequest()` should change?)
- Its based on the assumptions that its state will be changed by the abstract classes (e.g. you implicitly assume that will change the `attributes` field. otherwise it will fail)
- It's hard to test, because you need to test all via its subclasses

I think that you could fix most of these issues by using composition instead of inheritance. Plus use more functional style (a method should return something rather modify the state).

Reply:
Thanks for the feedback. I have some comments below.

> - It's very generic, one public method that takes some number of some objects and returns something

The return type for an operation is already genericized and thus well-typed, good point on the operation arguments.

> - Its abstract methods do not have a clear, well-defined contract (if you were to implement a new subclass, how would you know what prepareHttpRequest() should change?)

Not sure – `prepareHttpRequest()` and `unmarhal()` are the methods supposed to be implemented for each different AWS operation, the former to prepare a request from arguments, and the latter to build a return object from the server response. 

> - Its based on the assumptions that its state will be changed by the abstract classes (e.g. you implicitly assume that will change the attributes field. otherwise it will fail)

Hmm. Query attributes and headers are there purposely to be changed (i.e., prepared) in view of the REST API call to the AWS service. I don't get the "otherwise it will fail" comment.

> - It's hard to test, because you need to test all via its subclasses

Why? We do want to test the concrete subclasses that implement the different operations

> I think that you could fix most of these issues by using composition instead of inheritance. Plus use more functional style (a method should return something rather modify the state).

I believe inheritance is okay for this design and it's more a matter of taste than a real issue.

Please keep in mind that this hierarchy is an evolution of the previous design, so the final design is pragmatically shaped on keeping things 1) working throughout the refactoring 2) compatible with what we had before.

That said, I agree with making things more functional (I totally appreciate that) and refining the contract with subclasses.

Reply:
Reflecting upon your comments, I have refactored the code and switched to a different, cleaner design, so to separate the http client part from request data and response unmarshalling utility methods. Everything is more functional now and I got rid of the mutable state in operation invocation.